### PR TITLE
Fix embark-goto-location with previous next-error

### PR DESCRIPTION
**Pull request by aaronjensen**
In cases where `next-error` was used in another buffer prior to doing an
`embark-collect-snapshopt`, that buffer will be preferred when doing
`next-error`. By setting the `next-error-find-buffer-function`, we override that
behavior

Fixes the issue I mentioned here: https://github.com/oantolin/embark/issues/146#issuecomment-779989307

Another approach is to set `next-error-last-buffer`, but this is more direct and `next-error-find-buffer-function` has higher precedence.

### CODE REVIEW

**Comment by aaronjensen** [comment]
Also, I understand better now what this is doing and it's definitely inferior to a regular grep mode buffer. Specifically, because it uses a temp buffer with only a single line, one of `next-error`'s greatest features (imo) is unable to be leveraged. That is, you cannot just go to the next match with a key from any other buffer.

Ironically, this is the same feature that caused this issue in the first place.

So, I think I'll be just using export when I run a grep so I can get that behavior natively.

**Comment by minad** [comment]
@oantolin Go back to the old compile mode specific function instead? I guess this was the problem you had observed?

**Comment by oantolin** [comment]
> Also, I understand better now what this is doing and it's definitely inferior to a regular grep mode buffer. Specifically, because it uses a temp buffer with only a single line, one of `next-error`'s greatest features (imo) is unable to be leveraged. That is, you cannot just go to the next match with a key from any other buffer.
 
> So, I think I'll be just using export when I run a grep so I can get that behavior natively.

Absolutely, @aaronjensen! The embark collect buffers are not supposed to reimplement all the features of the corresponding embark export buffers, they are just supposed to be some least common denominator base functionality that works for every type of targets. It is expected than whenever there is an exporter for a particular type, the export buffers are more featureful than the collect buffers.

Basically collect gives functionality that can be made to work uniformly independently of the type of targets; while export gives access to functionality that is perfectly adapted to one specific type of targets.

I'll compare this patch to @minad's suggestion of returning to my old implementation and pick one or the other.

**Comment by aaronjensen** [comment]
Awesome, thanks.

It looks like the only way to detect if the current minibuffer command is a grep command is to actually attempt the export, is that right? I'm probably missing something here. I was trying to write something that'd detect if I'm in a grep and export, otherwise collect snapshot. I think actually I may just always export and see how that goes for a while.

**Comment by oantolin** [comment]
> It looks like the only way to detect if the current minibuffer command is a grep command is to actually attempt the export

You can test for the `xref-location` category, which is the one the grep exporter is used for. Use this expression:
`(eq (completion-metadata-get 'category (embark--metadata)) 'xref-location)`

> I think actually I may just always export and see how that goes for a while.

Export falls back to collect if no specific exporter is available, so I'd actually recommend this if you always want the maximum functionality. Of course, if you, for example, don't like dired for some reason, and never want to use it, you can just remove the dired exporter from `embark-exporters-alist`.

**Comment by oantolin** [comment]
I decided to go with the old implementation using `compile-goto-error`, since it seems more direct. 9b9aaff174a55eba76eb1b199dfdc4a52400c91e Please test, @aaronjensen.

**Comment by aaronjensen** [comment]
Hm, it works, but it opens two windows when I press enter for some reason.

**Comment by oantolin** [comment]
I saw the two window problem! But when I restarted Emacs to debug from a clean state, I couldn't reproduce the problem any more. I still can't reproduce it. So, maybe restart Emacs? It might also cure the problem for you. I know that's terrible advice, and I should figure out what caused it instead, but since I can't reproduce it figuring it out seems difficult.

**Comment by aaronjensen** [comment]
It reproduces for me after a restart

**Comment by aaronjensen** [comment]
Yeah, I can reproduce it pretty consistently. The embark snapshot window cannot be the only window. It must be the right window, I believe.

**Comment by oantolin** [comment]
OK, thanks for the tip, I'll look into it.

**Comment by oantolin** [comment]
Yep, that reproduces the problem, @aaronjensen! I'll see if I can figure it out.

**Comment by oantolin** [comment]
I think this does it. 040eb317bc422477ddb4eeff4a565d1e3c9024b6

**Comment by aaronjensen** [comment]
Hmm, I can still reproduce it in the same way even with that commit.

**Comment by oantolin** [comment]
Oh, I really thought I had it, @aaronjensen. Here's what I saw happen with my config: first, before that 040eb31 commit, if I followed your recipe, that is, if I opened two windows, ran `consult-grep` and then `embark-collect-snapshot`, and then pressed `RET` in the collect buffer, I would see the other window split in two and *both* show the result I pressed `RET` on. After that commit, however, I see the non-collect window split in two, only *one* of which shows the result I pressed `RET` on. And I think that's correct behavior, no?

You still see both windows showing the search result?

**Comment by aaronjensen** [comment]
Yes, I see both windows show the search result, but only if my emacs window is big enough... if it's small, it doesn't do it it seems. In the attached gif, I first do it with a smaller emacs window and then I do it with a full size (on my 16" laptop) window.


CleanShot 2021-02-20 at 20 25 17

**Comment by oantolin** [comment]
Thanks for the gif. At some point in the past, `embark-goto-location` always displayed it's search result in the target window, which seemed (a) consistent and predictable, (b) sensible, since usually the target window is incidental to a command that searches many files. I think that was @minad's idea, but I removed it, probably without much thought. I brought that feature back now, and hope that finally takes care of this problem. c23bb53054baab8b927eae7d818d446b39a8e92c

**Comment by aaronjensen** [comment]
Looks like it's working for me now, thanks!

**Comment by oantolin** [comment]
> Looks like it's working for me now, thanks!

Great! And thanks for your patience, @aaronjensen.
